fix(gossipsub): gracefully disable handler on stream errors

vnermolaev opened the pull request:
## Description

Previously, we closed the entire connection upon receiving too many upgrade errors. This is unnecessarily aggressive. For example, an upgrade error may be caused by the remote dropping a stream during the initial handshake which is completely isolated from other protocols running on the same connection.

Instead of closing the connection, set `KeepAlive::No`.

Related: #3591.
Resolves: #3690.

## Notes & open questions

Few items I'd like to raise.
- Pieces of code such as 
   ```rust
   return Poll::Ready(ConnectionHandlerEvent::Close(...));
   ```
   are replaced with `self.keep_alive = KeepAlive::No` and the error report. Unfortunately, I cannot meaningfully return from the `poll` function because it returns `ConnectionHamdlerEvent` and not an `Option` thereof.
- Pieces of code such as 
   ```rust
   loop {
       ...
       return Poll::Ready(ConnectionHandlerEvent::Close(...));
       ...
   }
   ```
   are replaced with `self.keep_alive = KeepAlive::No` and the error report. Again, it is impossible to return meaningfully; I just `break`; such an approach is also observed in the `poll` function.
   - sometimes errors are reported on the `warn!` level which is inconsistent with the proposed `info!` level

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

Commit:
Gossipsub: remove `ConnectionHandlerEvent::Close`

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -325,7 +325,8 @@ impl ConnectionHandler for Handler {
             if let Some(error) = reported_error {
-                return Poll::Ready(ConnectionHandlerEvent::Close(error));
+                self.keep_alive = KeepAlive::No;
+                log::info!("Gossipsub error: {error}");
```
I think this indirection is confusing.

We should "handle" the error as soon as possible. Now that we are changing the error handling from returning it to logging, we can do that as soon as the error occurs instead of storing it in a variable (`upgrade_errors`) and delaying the handling.

For the specific state where the protocol is unsupported, we can store only that state and return the specific event within `poll`.

Those that make sense?

vnermolaev replied:
Clear. FYI there are a few places where `reported_error` is set.

Still, it bugs me that there is no return statement. Because further processing seems like extra unnecessary work. This statement applies to all places where `return` is replaced by a report.

thomaseizinger replied:
> Still, it bugs me that there is no return statement. Because further processing seems like extra unnecessary work.

Yes, we shouldn't even be doing any of this if we can just log the error in the callback where we are first encountering it.

The original reason why this variable exists is because we can't do anything in the callback where it happens.

`ConnectionHandler` is a state machine and `poll` "advances" it. Thus, if errors happen in other parts we needed to sort-of buffer them and return them on the next "tick" (i.e. `poll`).

But now that we no longer want to return them (because returning would close the connection), we can get rid of this variable and "buffering".

vnermolaev replied:
A decision must be taken on how to return. Please see my other comment [below](https://github.com/libp2p/rust-libp2p/pull/3625#discussion_r1140067178)

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -475,14 +474,14 @@ impl ConnectionHandler for Handler {
-                                    error!("Error sending message: {}", e);
-                                    return Poll::Ready(ConnectionHandlerEvent::Close(e));
+                                    error!("Error sending message: {e}");
+                                    break;
```
I need to think about this in more detail and will come back with something!

vnermolaev replied:
My 5c:
- return `Poll<Option<...>>`. Breaking. Non-viable.
- return `Poll::Ready(ConnectionHandlerEvent::Custom(X))`, not clear what `X` is.
- introduce `NoOp` in `ConnectionHandlerEvent` (to mimic `Option`). Least invasive but adds a variant while another is being removed. 
- an option from a more informed person e.g., you :)

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -291,13 +290,17 @@ impl ConnectionHandler for Handler {
     > {
         // Handle any upgrade errors
         if let Some(error) = self.upgrade_errors.pop_front() {
```
Can't we just get rid of this entire field and handle the errors within the callback where they originate?

Commit:
Move error handling closer to the source.

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -340,9 +305,8 @@ impl ConnectionHandler for Handler {
-                HandlerError::MaxInboundSubstreams,
-            ));
+            self.keep_alive = KeepAlive::No;
+            log::info!("Gossipsub error: {}", HandlerError::MaxInboundSubstreams);
```
Lets simplify this to print the entire error and deprecate the variant.

vnermolaev replied:
addressed

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -351,9 +315,8 @@ impl ConnectionHandler for Handler {
-                    HandlerError::MaxOutboundSubstreams,
-                ));
+                self.keep_alive = KeepAlive::No;
+                log::info!("Gossipsub error: {}", HandlerError::MaxInboundSubstreams);
```
Here too. Note that this is also the wrong variant.

vnermolaev replied:
addressed

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -351,9 +315,8 @@ impl ConnectionHandler for Handler {
             && !self.outbound_substream_establishing
         {
             if self.outbound_substreams_created >= MAX_SUBSTREAM_CREATION {
```
In how many places do we increment `outbound_substreams_created`? Perhaps we can do this earlier as well?

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -475,14 +438,14 @@ impl ConnectionHandler for Handler {
                                 Err(e) => {
-                                    error!("Error sending message: {}", e);
-                                    return Poll::Ready(ConnectionHandlerEvent::Close(e));
+                                    error!("Error sending message: {e}");
```
`error` is too high a level, let's log these on debug.

vnermolaev replied:
this level has been set before I started working on this PR. So I am not sure which `error`s must be converted to `debug`s. I am doing it for this one.

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -475,14 +438,14 @@ impl ConnectionHandler for Handler {
                                 Err(e) => {
-                                    error!("Error sending message: {}", e);
-                                    return Poll::Ready(ConnectionHandlerEvent::Close(e));
+                                    error!("Error sending message: {e}");
```
I think we also need to set `outbound_substream` to `None` here otherwise we remain in "Poisoned".

vnermolaev replied:
We can end up with a `Poisoned` stream in multiple places which I have not modified. Just look for the `error`-level logs, `self.outbound_stream` is not set to `None` there.
E.g.
```rust
Poll::Ready(Err(e)) => {
    error!("Outbound substream error while sending output: {e:?}");
    break;
}
```                       
 Do you think I should also update it there?

thomaseizinger replied:
Previously, we returned `ConnectionHandlerEvent::Close` which meant we would never call `poll` again on this handler. This is changing now so we need to update the code accordingly so those state machines are never in poisoned.

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -504,7 +467,8 @@ impl ConnectionHandler for Handler {
                         }
                         Poll::Ready(Err(e)) => {
-                            return Poll::Ready(ConnectionHandlerEvent::Close(e))
+                            log::info!("NegotaitedSubstream error: {e}");
```
Can we make these according to the function that failed? In this case "flushing". Like "failed to flush outbound stream".

vnermolaev replied:
addressed

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -571,10 +529,37 @@ impl ConnectionHandler for Handler {
+                    // Timeout errors get mapped to NegotiationTimeout and we close the connection.
+                    ConnectionHandlerUpgrErr::Timeout | ConnectionHandlerUpgrErr::Timer => {
+                        self.keep_alive = KeepAlive::No;
+                        log::info!("Dial upgrade error: {}", HandlerError::NegotiationTimeout);
```
Here too, no need for the `HandlerError` indirection if we can just log it directly.

thomaseizinger replied:
I think we can deprecate the entire `HandlerError` type.

vnermolaev replied:
Unfortunately, no. Variants `MaxTransmissionSize` and `Codec` are used when implementing, e.g., `impl<TSocket> InboundUpgrade<TSocket> for ProtocolConfig`.
I have removed other variants.

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -571,10 +529,37 @@ impl ConnectionHandler for Handler {
+                    }
+                    // There was an error post negotiation, close the connection.
+                    ConnectionHandlerUpgrErr::Upgrade(UpgradeError::Apply(e)) => {
+                        self.keep_alive = KeepAlive::No;
```
This isn't fatal, lets not set `keep_alive` here but just log the error.

vnermolaev replied:
addressed

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -571,10 +529,37 @@ impl ConnectionHandler for Handler {
+                                log::info!("Dial upgrade error: {}", NegotiationError::Failed);
+                            }
+                            NegotiationError::ProtocolError(e) => {
+                                self.keep_alive = KeepAlive::No;
```
Same here, lets not set this `No`.

vnermolaev replied:
addressed

thomaseizinger reviewed (commented):
Nice, we are making progress in the right direction here.

vnermolaev commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -340,9 +305,8 @@ impl ConnectionHandler for Handler {
 
         if self.inbound_substreams_created > MAX_SUBSTREAM_CREATION {
```
Conditions for inbound and outbound streams vary, i.e.,
`self.inbound_substreams_created > MAX_SUBSTREAM_CREATION` vs
`self.outbound_substreams_created >= MAX_SUBSTREAM_CREATION`.
I am sticking to the latter option.

thomaseizinger replied:
I am almost certain this is an oversight so I am in favor of harmonizing.

cc @AgeManning.

AgeManning replied:
Yeah, looks like an oversight to me

Commit:
Address PR comments.

thomaseizinger commented on protocols/gossipsub/src/error_priv.rs:
```diff
@@ -89,16 +88,8 @@ impl From<SigningError> for PublishError {
-    #[error("Protocol negotiation timeout.")]
-    NegotiationTimeout,
-    #[error("Protocol negotiation failed.")]
-    NegotiationProtocolError(ProtocolError),
```
Can we keep these but mark as `deprecated` with a note that it will no longer be emitted? That would allow us to ship it as a patch-release instead of a breaking change.

vnermolaev replied:
addressed

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -536,11 +545,10 @@ impl ConnectionHandler for Handler {
                     // Timeout errors get mapped to NegotiationTimeout and we close the connection.
                     ConnectionHandlerUpgrErr::Timeout | ConnectionHandlerUpgrErr::Timer => {
                         self.keep_alive = KeepAlive::No;
```
One timeout is not fatal, let's remove this line please.

vnermolaev replied:
done

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -181,6 +181,13 @@ impl Handler {
+            self.keep_alive = KeepAlive::No;
+            log::info!("Gossipsub error: The maximum number of inbound substreams created has been exceeded.");
+            return;
+        }
```
Say we continuously fail to establish inbound streams. This function is only called upon successfully establishing an inbound stream. Thus, we would never set `keep_alive` to `No`, correct?

I think this check (and the one below) needs to be in `on_connection_event`. For `on_fully_negotiated_inbound` and `listen_error` we check the `inbound_substream_created` variable and the equivalent for the other.

Thoughts?

vnermolaev replied:
Functionally speaking, it only makes a little difference if you look at the code.
However, semantically, these checks are better placed in `on_connection even` indeed.
Moved.

Also, I did not manage to find a place where the counter is _de_ cremented.

thomaseizinger replied:
> Functionally speaking, it only makes a little difference if you look at the code.

I don't think I agree. Perhaps the name could be improved but if I understand correctly, the idea here was that we at most _try to create_ `MAX_SUBSTREAM_CREATION` substream. Thus we should check the counter against this upper bound for every successful **and** failed creation of a substream. See the doc comment on the constant.

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -181,6 +181,13 @@ impl Handler {
+        if self.inbound_substreams_created == MAX_SUBSTREAM_CREATION {
+            // Too many inbound substreams have been created, end the connection.
+            self.keep_alive = KeepAlive::No;
+            log::info!("Gossipsub error: The maximum number of inbound substreams created has been exceeded.");
```
```suggestion
            log::info!("The maximum number of inbound substreams created has been exceeded.");
```

Nit: This is already part of the module that is inferred from where we use the macro.

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -571,10 +538,32 @@ impl ConnectionHandler for Handler {
+                            NegotiationError::ProtocolError(e) => {
+                                log::info!("Gossipsub error: Protocol negotiation failed. {e}");
+                            }
+                        }
```
You can merge this match statement into the outer one and thus reduce one level of indentation.

vnermolaev replied:
done

thomaseizinger reviewed (commented):
Thanks!

Some more thoughts.

Commit:
Address PR comments.

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -475,14 +431,17 @@ impl ConnectionHandler for Handler {
                         Poll::Ready(Err(e)) => {
-                            error!("Outbound substream error while sending output: {:?}", e);
-                            return Poll::Ready(ConnectionHandlerEvent::Close(e));
+                            error!("Outbound substream error while sending output: {e:?}");
```
Minor consistency alignment.

```suggestion
                            log::debug!("Outbound substream error while sending output: {e}");
```

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -525,14 +485,8 @@ impl ConnectionHandler for Handler {
-                                )
-                                .into(),
-                            ));
+                            warn!("Outbound substream error while closing: {e:?}");
```
Same here.

```suggestion
                            log::debug!("Outbound substream error while closing: {e}");
```

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -566,15 +520,51 @@ impl ConnectionHandler for Handler {
+                match error {
+                    // Timeout errors get mapped to NegotiationTimeout and we close the connection.
+                    ConnectionHandlerUpgrErr::Timeout | ConnectionHandlerUpgrErr::Timer => {
+                        log::info!("Dial upgrade error: Protocol negotiation timeout.");
```
```suggestion
                        log::debug!("Dial upgrade error: Protocol negotiation timeout.");
```

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -566,15 +520,51 @@ impl ConnectionHandler for Handler {
+                        log::info!("Dial upgrade error: Protocol negotiation timeout.");
+                    }
+                    // There was an error post negotiation, close the connection.
+                    ConnectionHandlerUpgrErr::Upgrade(UpgradeError::Apply(e)) => {
```
This `e` is a `HandlerError` but effectively, the `InboundUpgrade` and `OutboundUpgrade` functions never fail. Can we replace these with `Void` please and use `void::unreachable(e)` here please?

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -566,15 +520,51 @@ impl ConnectionHandler for Handler {
+                    }
+                    ConnectionHandlerUpgrErr::Upgrade(UpgradeError::Select(
+                        NegotiationError::ProtocolError(e),
+                    )) => log::info!("Protocol negotiation failed. {e}"),
```
```suggestion
                    )) => log::debug!("Protocol negotiation failed: {e}"),
```

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -566,15 +520,51 @@ impl ConnectionHandler for Handler {
+                    )) => {
+                        // The protocol is not supported
+                        self.protocol_unsupported = true;
+                        log::info!("Dial upgrade error: {}", NegotiationError::Failed);
```
```suggestion
                        log::debug!("The remote peer does not support gossipsub on this connection");
```

thomaseizinger commented on protocols/gossipsub/src/error_priv.rs:
```diff
@@ -89,14 +89,18 @@ impl From<SigningError> for PublishError {
 /// Errors that can occur in the protocols handler.
 #[derive(Debug, Error)]
 pub enum HandlerError {
+    #[deprecated(note = "This error will not be emitted")]
```
```suggestion
    #[deprecated(note = "This error will no longer be emitted")]
```

thomaseizinger reviewed (commented):
Thanks for pushing this forward!

Commit:
Address PR comments.

thomaseizinger committed:
Use `void` instead of panic

thomaseizinger committed:
Check created streams counter also for failed upgrades

thomaseizinger reviewed (approved):
This is good from my end. I would like @mxinden to also have a look.

vnermolaev committed:
Merge branch 'libp2p:master' into deprecate/gossipsub-close-event

Commit:
Update changelog.

thomaseizinger commented on protocols/gossipsub/CHANGELOG.md:
```diff
@@ -2,8 +2,11 @@
+  See [PR 3625].
 
-[PR 3551]: https://github.com/libp2p/rust-libp2p/pull/3551
+[PR 3551]: https://github.com/libp2p/rust-libp2p/pull/3551git pull
```
```suggestion
[PR 3551]: https://github.com/libp2p/rust-libp2p/pull/3551
```

vnermolaev replied:
yes, this was an accidental focus switch :(

thomaseizinger committed:
Fix typo in changelog

mxinden commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -564,17 +517,65 @@ impl ConnectionHandler for Handler {
-                self.upgrade_errors.push_back(e);
+
+                match error {
+                    // Timeout errors get mapped to NegotiationTimeout and we close the connection.
```
Is this still true?

thomaseizinger replied:
Seems outdated, nice catch!

mxinden commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -564,17 +517,65 @@ impl ConnectionHandler for Handler {
+                    ConnectionHandlerUpgrErr::Timeout | ConnectionHandlerUpgrErr::Timer => {
+                        log::debug!("Dial upgrade error: Protocol negotiation timeout.");
+                    }
+                    // There was an error post negotiation, close the connection.
```
Is this still true?

thomaseizinger replied:
Seems outdated, nice catch!

mxinden reviewed (commented):
Overall change looks good to me. Thanks for the work.

//CC @divagant-martian and @AgeManning to make sure this does not break any assumptions in lighthouse.

thomaseizinger committed:
Remove outdated comments

thomaseizinger committed:
Extract utility functions for classifying `ConnectionEvent`

thomaseizinger committed:
Set `outbound_substream_establishing` in a single place

thomaseizinger committed:
Flatten match

mxinden reviewed (approved):
Great. Thanks for the follow-ups.

Let's give the sigma prime folks a bit more time to comment.

AgeManning commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -470,19 +424,22 @@ impl ConnectionHandler for Handler {
-                            error!("Outbound substream error while sending output: {:?}", e);
-                            return Poll::Ready(ConnectionHandlerEvent::Close(e));
+                            log::debug!("Outbound substream error while sending output: {e}");
+                            break;
```
I think this leaves the Substream state in the Poisoned state, which is going to cause gossipsub to panic on line 502

AgeManning commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -504,7 +461,8 @@ impl ConnectionHandler for Handler {
                         Poll::Ready(Err(e)) => {
-                            return Poll::Ready(ConnectionHandlerEvent::Close(e))
+                            log::debug!("Outbound substream error while flushing output: {e}");
+                            break;
```
Similarly here.

AgeManning commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -525,14 +483,8 @@ impl ConnectionHandler for Handler {
-                                .into(),
-                            ));
+                            log::debug!("Outbound substream error while closing: {e}");
+                            break;
```
And here

AgeManning commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -564,17 +516,58 @@ impl ConnectionHandler for Handler {
+        if event.is_inbound() && self.inbound_substreams_created == MAX_SUBSTREAM_CREATION {
+            // Too many inbound substreams have been created, disable the handler.
+            self.keep_alive = KeepAlive::No;
+            log::info!("The maximum number of inbound substreams created has been exceeded.");
```
Maybe this should be a warn, error or a crit. In principle this should never happen and we don't want the connection to just end silently

thomaseizinger replied:
The thinking was that `info!` is typically used for a state change (we are disabling the handler). I don't want to wake an ops person at 3am because we are emitting `warn` or `error` here :)

Happy to be convinced otherwise.

AgeManning replied:
Yeah, I'm not all that concerned. I;d be curious to know when to use a warn then. I'd typically use a warn if something has not behaved as it should and is worth notifying a user. 
In this case, we pretty much should never see this log, unless something is pretty broken. I find it handy to grep warn/error etc to find things that are broken. This fits into that classification imo. 

But if other parts of libp2p are not doing it, happy to leave as an info.

thomaseizinger replied:
Why would we never see this log? We would see this if we connect to a node that doesn't support the gossipsub protocol right? Nothing is inherently broken if that happens so I am not sure a `warn` is appropriate.

mxinden replied:
Agreed that this should rather be a `warn` instead of a `info`.

> Why would we never see this log?

By default `env-logger` does not log on `info` level. I would argue that users should see this by default, thus it should be a level higher than `info` which is `warn`.

> In this case, we pretty much should never see this log, unless something is pretty broken. I find it handy to grep warn/error etc to find things that are broken. This fits into that classification imo.

Agreed.

> We would see this if we connect to a node that doesn't support the gossipsub protocol right?

As far as I can tell a `UpgradeError::Select(NegotiationError::Failed)` would lead to `self.protocol_unsupported = true` and thus a direct disabling (i.e. `self.keep_alive = KeepAlive::No`) in the next `ConnectionHandler::poll` invocation. Thus we would never request another outbound stream and thus the log line above is never printed. Am I missing something @thomaseizinger?

Just to avoid confusion, I am in favor of the above behavior. I don't think there is value in directly requesting another stream in case the remote signaled that it does not support the protocol on the previous stream. Objections?

thomaseizinger replied:
Makes sense!

AgeManning commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -564,17 +516,58 @@ impl ConnectionHandler for Handler {
+                error: ConnectionHandlerUpgrErr::Timeout | ConnectionHandlerUpgrErr::Timer,
+                ..
+            }) => {
+                log::debug!("Dial upgrade error: Protocol negotiation timeout.");
```
We only do the negotiation once to create long-lived substreams. If the negotiation times out, I rarely find it useful to continually re-connect and retry the negotiations, which is why I previously opted to just close the connection. 

If I understand these changes, every message we want to send, we are going to continually try and re-negotiate?

thomaseizinger replied:
We can also consider this a "hard error" for gossipsub and immediately disable the handler.

thomaseizinger replied:
Currently, we'd try `MAX_SUBSTREAM_CREATION` times and then disable the handler.

AgeManning replied:
Yep, that seems like reasonable logic. 

If there were another behaviour that had a keepalive::Yes. It would continue to try and reconnect tho right?

mxinden replied:
> Currently, we'd try `MAX_SUBSTREAM_CREATION` times and then disable the handler.

I don't think this is correct. We increase `outbound_substream_established` only on successful substream negotiation (i.e. in `on_fully_negotiated_outbound`). The above failure does not increase the counter and thus the limit `MAX_SUBSTREAM_CREATION` is never hit.

thomaseizinger replied:
> > Currently, we'd try `MAX_SUBSTREAM_CREATION` times and then disable the handler.
> 
> I don't think this is correct. We increase `outbound_substream_established` only on successful substream negotiation (i.e. in `on_fully_negotiated_outbound`). The above failure does not increase the counter and thus the limit `MAX_SUBSTREAM_CREATION` is never hit.

You are totally right, not sure how I missed that. Do we agree that the logic I described is what we want to have?

mxinden replied:
> Do we agree that the logic I described is what we want to have?

Agreed in case it includes the re-enqueuing of the RPC of the failed outbound stream. See https://github.com/libp2p/rust-libp2p/pull/3625#discussion_r1150749684.

thomaseizinger replied:
Yep I think that makes sense.

mxinden replied:
Tracking the number of requested outbound streams instead of the number of successfully upgraded outbound streams is done with https://github.com/libp2p/rust-libp2p/pull/3625/commits/0507493a1e5232047c8495319e82b00bd80303eb.

AgeManning commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -564,17 +516,58 @@ impl ConnectionHandler for Handler {
+            ConnectionEvent::DialUpgradeError(DialUpgradeError {
+                error: ConnectionHandlerUpgrErr::Upgrade(UpgradeError::Apply(e)),
+                ..
+            }) => void::unreachable(e),
```
Is this for sure unreachable?

thomaseizinger replied:
Yes, proven by the type-system because e is `Void`.

AgeManning commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -564,17 +516,58 @@ impl ConnectionHandler for Handler {
+            self.outbound_substream_establishing = false;
+        }
+
+        if event.is_inbound() && self.inbound_substreams_created == MAX_SUBSTREAM_CREATION {
```
I typically prefer to use `>=` than an `==`. 
Although it shouldn't matter in this current instance, it can be more robust against modifications in the future, for example, someone incrementing inbound_substreams_created in the `poll()`, which causes us to exceed the limit.

AgeManning commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -564,17 +516,58 @@ impl ConnectionHandler for Handler {
+        if event.is_outbound() && self.outbound_substreams_created == MAX_SUBSTREAM_CREATION {
+            // Too many outbound substreams have been created, disable the handler.
+            self.keep_alive = KeepAlive::No;
+            log::info!("The maximum number of outbound substreams created has been exceeded.");
```
Similarly here, I think this should be a warn, error, or crit.

AgeManning commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -564,17 +516,58 @@ impl ConnectionHandler for Handler {
+                    )),
+                ..
+            }) => {
+                log::debug!("Protocol negotiation failed: {e}")
```
Similarly here. ARe we going to constantly try and re-negotiate? I think we should either drop the connection of set KeepAlive to No.

thomaseizinger replied:
Up to `MAX_SUBSTREAM_CREATION` times only, then the handler will be set to `KeepAlive::No`.

AgeManning reviewed (changes requested):
I think we need to be careful when changing the handler, because it can be pretty hard to debug problems we may introduce and can run into panics fairly easily. 

I think this should be tested pretty thoroughly before putting it into a release. 

I haven't really thought through all the intermediate states we could get in when things go wrong but we leave the connection alive. 

It looks like at a first glance some things are missed which currently will cause the code to panic.

thomaseizinger committed:
Avoid being stuck in "Poisoned" state for outbound streams

thomaseizinger committed:
Don't handle error that is never constructed

thomaseizinger committed:
Deprecate `HandlerError` entirely

AgeManning reviewed (dismissed):
Look ok to me now. I think @divagant-martian may be having a look over too soon.

mxinden commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -564,17 +505,58 @@ impl ConnectionHandler for Handler {
-                warn!("Dial upgrade error {:?}", e);
-                self.upgrade_errors.push_back(e);
+            ConnectionEvent::DialUpgradeError(DialUpgradeError {
+                error: ConnectionHandlerUpgrErr::Timeout | ConnectionHandlerUpgrErr::Timer,
```
In case we stick with retrying after an outbound substream failed.

```suggestion
                error: ConnectionHandlerUpgrErr::Timeout | ConnectionHandlerUpgrErr::Timer,
                info,
```

I suggest requeuing the RPC into the front of `self.send_queue.push_front(info)`. Otherwise the remote may receive a `GRAFT` before a `SUBSCRIBE`. Related to https://github.com/libp2p/rust-libp2p/issues/3690.

mxinden replied:
Re-enqueuing is done with https://github.com/libp2p/rust-libp2p/pull/3625/commits/b5728952464f570c5636f5e6a4fa2448e28f3cbe.

mxinden committed:
Track # of outbound streams requested, not successfully established

The GossipSub protocol uses two streams, namely one from the local to the remote and one from the
remote to the local node. The rust-libp2p implementation enforces an upper limit on the churn of
both via `MAX_SUBSTREAM_CREATION`.

Thus far `self.outbound_substreams_created` was increased for every successful stream, but ignored
failed streams. Thus, on a given connection, a GossipSub handler could potentially indefinitely
retry creating an outbound stream, where each of those streams fails instead of succeeds to upgrade.

With this commit the handler tracks the number of requested outbound streams instead.

mxinden committed:
Re-enqueue message when outbound stream fails

An outbound stream carries the first message for the remote node when upgrading in the
`OutboundOpenInfo`. Thus far the GossipSub handler would drop the message when the outbound stream
fails to upgrade.

This could e.g. lead to a remote peer receiving a GRAFT without a previous SUBSCRIBE.

With this commit the message carried in the upgrade's `OutboundOpenInfo` is re-enqueued.

thomaseizinger committed:
Use early return instead of if-else

thomaseizinger committed:
Only send messages in `poll`

Instead of popping messages early and having to re-queue them on error,
we can remove the use of `OutboundOpenInfo` and only send messages
in one place.

thomaseizinger commented on swarm/src/handler.rs:
```diff
@@ -212,6 +212,31 @@ pub enum ConnectionEvent<'a, IP: InboundUpgradeSend, OP: OutboundUpgradeSend, IO
+        matches!(
+            self,
+            Self::FullyNegotiatedInbound(_)
+                | Self::DialUpgradeError(DialUpgradeError {
```
Whoops, this needs to be `ListenUpgradeError`!

thomaseizinger reviewed (commented):
Found one bug at least.

thomaseizinger committed:
Fix use of `DialUpgradeError` in `is_inbound` check

mxinden commented on protocols/gossipsub/CHANGELOG.md:
```diff
@@ -2,8 +2,11 @@
 
 - Signed messages now use sequential integers in the sequence number field.
   See [PR 3551].
+- Gracefully disable handler on stream errors. Deprecate a few variants of `HandlerError`.
+  See [PR 3625].
```
Note for ourselves that this needs to move to a new `v0.44.3` section.

thomaseizinger replied:
Done.

thomaseizinger committed:
Merge branch 'master' into deprecate/gossipsub-close-event

thomaseizinger committed:
Move changelog entry

thomaseizinger committed:
Bump version

thomaseizinger committed:
Change log level to warn for bad events

Hitting the max substream creation means something bad is happening.
This is worth reporting as `warn`.

thomaseizinger committed:
Don't end log messages with periods

divagant-martian commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -469,20 +406,19 @@ impl ConnectionHandler for Handler {
                                 Err(e) => {
-                                    error!("Error sending message: {}", e);
-                                    return Poll::Ready(ConnectionHandlerEvent::Close(e));
+                                    log::debug!(
```
The logs for errors on outbound substreams have all been downgraded to `debug` while the inbound ones are kept as `warn`. I would expect the `pub` part of gossipsub to be as important (if not more) as the `sub` part to keep these all as `warn` at least

thomaseizinger replied:
To me, a `warn` always means "Should I wake an ops person at 3am because of this"? Connections can die at any time because somebody e.g. closes their laptop. That is not a reason to wake an ops person IMO, hence I am gonna downgrade the inbound streams to debug instead upgrading the outbound ones to warn.

divagant-martian commented on swarm/src/handler.rs:
```diff
@@ -212,6 +212,31 @@ pub enum ConnectionEvent<'a, IP: InboundUpgradeSend, OP: OutboundUpgradeSend, IO
+                    ..
+                })
+        )
+    }
```
Just a minor note here. While `ConnectionEvent` probably has changes not too often, this will be easily missed to be updated if a new variant is added. Personal preference would be to include something closer to a full match

thomaseizinger replied:
Fair!

thomaseizinger committed:
Use exhaustive match

thomaseizinger committed:
Make error message consistently `debug` and use same wording

thomaseizinger committed:
Merge branch 'master' into deprecate/gossipsub-close-event

mxinden commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -338,28 +295,17 @@ impl ConnectionHandler for Handler {
-                HandlerError::MaxInboundSubstreams,
-            ));
-        }
+        // Invariant: `self.inbound_substreams_created < MAX_SUBSTREAM_CREATION`.
```
How about adding an `assert!` or `debug_assert!` (preference for the former) here?

mxinden replied:
Taking a closer look, I don't think this invariant actually holds.

In `on_connection_event` we drop new inbound requests on `==`:

``` rust
        if event.is_inbound() && self.inbound_substreams_created == MAX_SUBSTREAM_CREATION {
            // Too many inbound substreams have been created, disable the handler.
            self.keep_alive = KeepAlive::No;
            log::warn!("The maximum number of inbound substreams created has been exceeded");
            return;
        }
```

Thus `self.inbound_substreams` may never be `>` but it could be `==` and thus the invariant of `<` does not hold.

@thomaseizinger am I missing something?

mxinden commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -381,25 +327,16 @@ impl ConnectionHandler for Handler {
+                            log::debug!("Failed to read from inbound stream: {error}");
+                            // Close this side of the stream. If the
+                            // peer is still around, they will re-establish their
+                            // connection
```
```suggestion
                            // peer is still around, they will re-establish their
                            // outbound stream i.e. our inbound stream.
```

It is not a connection that they reestablish, but a stream, right?

mxinden commented on swarm/src/handler.rs:
```diff
@@ -212,6 +212,38 @@ pub enum ConnectionEvent<'a, IP: InboundUpgradeSend, OP: OutboundUpgradeSend, IO
     ListenUpgradeError(ListenUpgradeError<IOI, IP>),
 }
 
+impl<'a, IP: InboundUpgradeSend, OP: OutboundUpgradeSend, IOI, OOI>
```
Note that this needs a `libp2p-swarm` version bump and a bump of the `libp2p-swarm` dependency in the `libp2p-gossipsub` `Cargo.toml`.

thomaseizinger replied:
Right, because we released since we started this :)

mxinden replied:
Done in bbdf8f5fb.

mxinden reviewed (approved):
3 more comments. Otherwise looks good to me.

mxinden committed:
Update protocols/gossipsub/src/handler.rs

mxinden committed:
chore: bump libp2p-swarm to v0.42.2

mxinden committed:
Update Cargo.lock

mxinden commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -564,17 +495,58 @@ impl ConnectionHandler for Handler {
+            self.keep_alive = KeepAlive::No;
+            log::warn!("The maximum number of outbound substreams created has been exceeded");
+            return;
+        }
```
As far as I can tell there are still multiple issues with this code.

- In case the remote node joins the local nodes mesh (`HandlerIn::JoinedMesh`), `self.keep_alive = KeepAlive::No` will be replaced by `KeepAlive::Yes`, see `on_behaviour_event`. Also in case the remote node leaves the local nodes mesh, `KeepAlive::Until` is set. Though the `ConnectionHandler` will still not be able to establish an outbound stream as `MAX_SUBSTREAM_CREATION` is hit.
- We may still emit `OutboundSubstreamRequest` in `ConnectionHandler::poll` as it does not check `self.outbound_substreams_requested <= MAX_SUBSTREAM_CREATION` first. That leads to wasted stream upgrades as that stream will then be dropped here.

I will have to think about this with a clear mind, i.e. tomorrow. Maybe we should go @thomaseizinger suggested route of modelling all of these states (keep alive, disabling, not supported, ...) as a combined `enum` to reduce the amount of possible state variants at compile time.

mxinden replied:
See https://github.com/libp2p/rust-libp2p/pull/3625#issuecomment-1498924751 for a potential solution.

mxinden committed:
Refactor keep alive mechanism

- Split handler into `EnabledHandler` and `DisabledHandler`, reducing the amount of potential state
variants (e.g. protocol-not-supported but keep-alive-yes) and enforce the possible states at compile time.

- Return `DeniedUpgrade` when handler is disabled (i.e. `DisabledHandler`).

- Don't update `KeepAlive` across all handler methods, but instead derive it ad-hoc in
`ConnectionHandler::connection_keep_alive()` based on the state of the handler. Again reduces number
of possible invalid states at compile time.

- Remove INITIAL_KEEP_ALIVE. The default idle timeout is 120s. This should be plenty enough time to
establish in and outbound streams.

mxinden committed:
Do minor clean up

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -485,6 +390,90 @@ impl ConnectionHandler for Handler {
+                    handler.in_mesh = false;
+                }
+            },
+            Handler::Disabled(_) => {}
```
Log a `debug` here that we are dropping the message?

mxinden replied:
:+1: b01e86f59

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -495,60 +484,83 @@ impl ConnectionHandler for Handler {
-                ..
-            }) => {
-                log::debug!("Protocol negotiation failed: {e}")
+                match event {
```
We started adding `transpose` functions to this event such that you can split it directly into an `Either` and can combine things like `Handler::Disabled` and the `unreachable` into one match arm. Might be worth doing here as well.

mxinden replied:
Neat! But if I am not mistaken the `transpose` methods are `pub(crate)`. Was that a deliberate choice? If not, mind pushing a commit to this branch @thomaseizinger?

thomaseizinger replied:
Ah interesting. I don't see a reason why they couldn't be public as well. Feel free to make them public.

mxinden replied:
`libp2p-swarm` implements `transpose` for `    FullyNegotiatedInbound<Either<_, _>>, Either<_, _>>`. Note the second `Either`. In this case both `ProtocolConfig` and `DeniedUpgrade` use a `()` for the `InboundUpgradeInfo`, thus the `libp2p-swarm` `transpose` does not apply.

@thomaseizinger I suggest we proceed here without this change. Fine by you?

thomaseizinger replied:
Yep, fine by me.

thomaseizinger commented on swarm/CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+## 0.42.2 - unreleased
+
+- Add `ConnectionEvent::` `is_outbound` and `is_inbound`. See [PR 3625].
```
```suggestion
- Add `ConnectionEvent::{is_outbound,is_inbound}`. See [PR 3625].
```

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -564,19 +486,81 @@ impl ConnectionHandler for Handler {
+                    );
+                    *self = Handler::Disabled(DisabledHandler::MaxStreamCreation);
+                    return;
+                }
```
If I understand correctly, than these counters are incremented elsewhere right?

For listening we likely increment them upon successful stream creations, for outbound, we increment them when we request the outbound stream. Two questions:

1. Should we account for failed inbound streams as well?
2. If yes to (1), should we increment these counters here as part of checking `is_inbound` and `is_outbound`?

mxinden replied:
Good idea. Done with 552cb0891.

mxinden committed:
Merge branch 'master' of https://github.com/libp2p/rust-libp2p into deprecate/gossipsub-close-event

thomaseizinger reviewed (commented):
Thank you Max!

mxinden committed:
Add debug for dropped message

mxinden committed:
Track MAX_SUBSTREAM_ATTEMPTS in on_connection_event

mxinden committed:
Update swarm/CHANGELOG.md

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -564,19 +485,83 @@ impl ConnectionHandler for Handler {
+
+                if event.is_outbound() {
+                    handler.outbound_substream_establishing = false;
+
```
```suggestion
```

thomaseizinger commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -564,19 +485,83 @@ impl ConnectionHandler for Handler {
+                    ConnectionEvent::FullyNegotiatedInbound(FullyNegotiatedInbound {
+                        protocol,
+                        ..
+                    }) => match protocol {
```
Could flatten this `match` into the outer one.

thomaseizinger reviewed (dismissed):
LGTM. Should we extract the changes around `void` and `ConnectionEvent` into separate PRs?

divagant-martian commented on protocols/gossipsub/src/handler.rs:
```diff
@@ -554,6 +389,92 @@ impl ConnectionHandler for Handler {
+
+                KeepAlive::Until(handler.last_io_activity + handler.idle_timeout)
+            }
+            Handler::Disabled(_) => KeepAlive::No,
```
I think this way it's possible we don't send the `PeerKind::UnsupportedProtocol` "on time" unless there is some kind of guarantee that poll will be called first **(?)**.  This event is important since we might want to get rid of (ban, or otherwise prevent connections to) this peer if it doesn't support what we need

thomaseizinger replied:
A handler is always completed drained of its events first before we even ask whether or not it should be shut down.

thomaseizinger replied:
See https://github.com/libp2p/rust-libp2p/blob/06ea8cef7ef6d8527fa9fdb9fc23dec88abe2a7b/swarm/src/connection.rs#L216-L303.

mxinden replied:
Adding to the above, we first "drain" (i.e. `poll`) the `ConnectionHandler` till it returns `Poll::Pending`:

https://github.com/libp2p/rust-libp2p/blob/06ea8cef7ef6d8527fa9fdb9fc23dec88abe2a7b/swarm/src/connection.rs#L231-L247

Only then do we check the `connection_keep_alive` statsu:

https://github.com/libp2p/rust-libp2p/blob/06ea8cef7ef6d8527fa9fdb9fc23dec88abe2a7b/swarm/src/connection.rs#L284-L303

mxinden committed:
Merge branch 'master' into deprecate/gossipsub-close-event

mxinden committed:
Merge branch 'master' into deprecate/gossipsub-close-event

divagant-martian reviewed (approved):
LGTM, management of handler state as an enum seems like a great code quality improvement on top, ty @mxinden

mxinden committed:
Merge branch 'master' into deprecate/gossipsub-close-event
